### PR TITLE
Adding functions to subject model

### LINKED ISSUE #34: Adding functionality to subject model
- Adding functionality to subject model
- Adding new file called test_subject.py for testing

### PR DESCRIPTION
**Pull request by yaelnoyman1**
Updating modles.py and create test_subject.py
Resolve #34

### CODE REVIEW

**Review by NitzanMordhai — CHANGES_REQUESTED** [review]
fix typo in the commit (modle.py --> models.py)

**Review by Shalevro2 — APPROVED** [review]
LGTM

**Comment by Yarboa** [comment]
> fix typo in the commit (modle.py --> models.py)

@yaelnoyman1 ^^

**Review by Yarboa — CHANGES_REQUESTED** [review]
please use fixtures in tests, for creating the data for the tests,

Consult with @Anas-mhameed look at his PR

File: projboard/tests/test_subject.py [hunk 1]
```diff
@@ -7,36 +7,43 @@
...
+    subject = Subject(name=name)
+    subject.save()
+    return subject
+
+
 class TestSubjectModel:
     @pytest.mark.django_db
     def test_create_subject(db):
-        subject = Subject.create_new_subject(name=SPORT)
+        subject = Subject.create_new_subject(SPORT)
```
Comment (Yarboa) [inline projboard/tests/test_subject.py]: Consider this please 
 @pytest.mark.django_db
@pytest.fixture
def generate_subject(name='name'):
    subject = Subject(name=name)
    subject.save()
    return subject
    
    
    It is called as following:
     def test_create_subject(generate_subject):
        '''Here you have to search in database the subject'''
        
        assert generate_subject.name==<to the returned subject from db>
        Please consult others, to understand how to write the tests

File: projboard/tests/test_subject.py [hunk 2]
```diff
@@ -7,36 +7,43 @@
...
         assert subject.name == SPORT
 
     @pytest.mark.django_db
-    def test_update_subject(db):
-        Subject.create_new_subject(name=SPORT)
-        subject = Subject.edit_subject(SPORT, ANIMALS)
+    def test_update_subject(self, generate_subject):
+        Subject.create_new_subject(generate_subject)
```
Comment (Yarboa) [inline projboard/tests/test_subject.py]: Please review test material again
Please review all tests

**Review by Yarboa — CHANGES_REQUESTED** [review]

**Review by rawadabu — APPROVED** [review]

File: projboard/models.py [hunk 3]
```diff
@@ -19,11 +19,68 @@ def get_subject_by_name(name):
...
+    def get_list_subjects_names():
+        """
+        Method get list subject names
+        :return: the list of subject
+        """
+        subjects_list = list(Subject.objects.all().values_list('name', flat=True))
+        return subjects_list
+
+    @staticmethod
+    def create_new_subject(name):
```
Comment (avi-biton) [inline projboard/models.py]: This method is not only creating a new subject, it also returns an existing subject (if exist).  
Consider renaming the method to `get_subject`

File: projboard/models.py [hunk 4]
```diff
@@ -19,11 +19,68 @@ def get_subject_by_name(name):
...
+        try:
+            subject = Subject.objects.get(name=name)
+            return subject
+        except Subject.DoesNotExist:
+            subject = Subject(name=name)
+            subject.save()
+            return subject
+
+    @staticmethod
+    def edit_subject(existing_name, new_name):
```
Comment (avi-biton) [inline projboard/models.py]: Consider renaming the method to `rename_subject` as it is not editing the subject, its just renaming it.

**Review by Shalevro2 — APPROVED** [review]
LGTM

File: projboard/models.py (hunk 3 above)
Comment (yaelnoyman1) [reply projboard/models.py]: I changed to "create_and_get_subject"

File: projboard/tests/test_subject.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    def test_update_subject(self, generate_subject):
+        subject = generate_subject[0]
+        subject = Subject.edit_subject(generate_subject[0], ANIMALS)
+        assert subject.name == ANIMALS
+
+        subject = Subject.edit_subject(NOTEXIST, ANIMALS)
+        assert not subject
+
+    @pytest.mark.django_db
+    def test_get_list_subjects_names(self, generate_subject):
```
Comment (Yarboa) [inline projboard/tests/test_subject.py]: Consider this
listSubject = Subject.get_list_subjects_names()
for i in generate_subject:
     assert subject.name in listSubject

File: projboard/tests/test_subject.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        subject = generate_subject[0]
+        assert subject.name == NAME
+
+    def test_delete_subject(self, generate_subject):
+        assert Subject.delete_subject(generate_subject[0])
+        subject = Subject.get_subject_by_name(generate_subject[0])
+        assert not subject
+
+    def test_update_subject(self, generate_subject):
+        subject = generate_subject[0]
```
Comment (Yarboa) [inline projboard/tests/test_subject.py]: Can you describe what you test here?

File: projboard/models.py (hunk 4 above)
Comment (Yarboa) [reply projboard/models.py]: Any updates to this?

**Review by Yarboa — CHANGES_REQUESTED** [review]

File: projboard/tests/test_subject.py (hunk 6 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py]: I now see that line 32 is not necessary
In line 33 I change the existing name to another name and in line 34 I check if it has changed

File: projboard/tests/test_subject.py (hunk 5 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py]: Done

File: projboard/models.py (hunk 4 above)
Comment (yaelnoyman1) [reply projboard/models.py]: Done

**Comment by Shalevro2** [comment]
Hi @yaelnoyman1,

I think you need to organize your code according #40, #30, #43, #41.
Take a look in the PR's above. We split models.py to directory that called models, according to this change you need also to change all the imports in the migrations files and tests.

**Comment by finziamit** [comment]
> Hi @yaelnoyman1,
> 
> I think you need to organize your code according #40, #30, #43, #41. Take a look in the PR's above. We split models.py to directory that called models, according to this change you need also to change all the imports in the migrations files and tests.

+1
Please separate the subject from models.py file to projboard/models/subject.py (also relocate the __init__.py file to models repository) and fix the imports

**Comment by Yarboa** [comment]
Please rebase main 
resolve conflict and push --force

**Comment by yaelnoyman1** [comment]
> Please rebase main resolve conflict and push --force

Done

File: projboard/tests/test_subject.py (hunk 1 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py]: I changed it to this, is that ok?

@pytest.fixture
@pytest.mark.django_db
def generate_subject():
    subject1 = Subject.create_and_get_subject(name="yael")
    subject2 = Subject.create_and_get_subject(name="gal")
    subject1.save()
    subject2.save()
    return [subject1, subject2]

File: projboard/tests/test_subject.py (hunk 2 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py]: ^^

File: projboard/tests/test_subject.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+ANIMALS = "Animals"
+NOTEXIST = "notexist"
+REDHAT = "redhat"
+NAME = "yael"
+
+
+@pytest.fixture
+@pytest.mark.django_db
+def generate_subject():
```
Comment (Yarboa) [inline projboard/tests/test_subject.py:12]: Please use two fixtures

If you need 1 use fixture for one if you need more, use another fixture, Please review all the tests

def generate_subject():
    subject = Subject.create_and_get_subject(name="yael")
    subject.save()
    return subject

and another one  to return more
def generate_subjects(generate_subject)
  subject = Subject.create_and_get_subject(name="yael")
    subject.save()
    return [subject, generate_subject]

**Review by Yarboa — CHANGES_REQUESTED** [review]

File: projboard/tests/test_subject.py (hunk 7 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py:12]: Done

**Comment by Anas-mhameed** [comment]
I suggest you squash all commits into just one commit

File: projboard/tests/test_subject.py (hunk 7 above)
Comment (Yarboa) [reply projboard/tests/test_subject.py:12]: @yaelnoyman1 I assume it is done
Please reply on the conversations

**Review by Yarboa — APPROVED** [review]

File: projboard/tests/test_subject.py (hunk 2 above)
Comment (Yarboa) [reply projboard/tests/test_subject.py]: @yaelnoyman1 I assume this is done
Correct?

File: projboard/models.py (hunk 4 above)
Comment (Yarboa) [reply projboard/models.py]: @yaelnoyman1 Please reply here

File: projboard/tests/test_subject.py (hunk 6 above)
Comment (Yarboa) [reply projboard/tests/test_subject.py]: Closing this

File: projboard/tests/test_subject.py (hunk 2 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py]: Yes

**Comment by yaelnoyman1** [comment]
> fix typo in the commit (modle.py --> models.py)

done

**Comment by yaelnoyman1** [comment]
> Hi @yaelnoyman1,
> 
> I think you need to organize your code according #40, #30, #43, #41. Take a look in the PR's above. We split models.py to directory that called models, according to this change you need also to change all the imports in the migrations files and tests.

Done, And I did rebase main

File: projboard/models.py (hunk 4 above)
Comment (yaelnoyman1) [reply projboard/models.py]: I changed the name to "rename_suject"

File: projboard/tests/test_subject.py (hunk 7 above)
Comment (yaelnoyman1) [reply projboard/tests/test_subject.py:12]: I changed the fixture like you asked
and changed the functions

**Pull request merged**